sys/kern/subr_uio.c

```diff
@@ -69,9 +69,68 @@
 SYSCTL_INT(_kern, KERN_IOV_MAX, iov_max, CTLFLAG_RD, SYSCTL_NULL_INT_PTR, UIO_MAXIOV,
 	"Maximum number of elements in an I/O vector; sysconf(_SC_IOV_MAX)");
 
+static uma_zone_t uio_zone;
+
 static int uiomove_flags(void *cp, int n, struct uio *uio, bool nofault,
     bool preserve_tags);
 
+struct uio *
+allocuio(u_int iovcnt)
+{
+	return uma_zalloc_arg(uio_zone, (void *)(uintptr_t)(iovcnt), M_WAITOK);
+}
+
+void
+freeuio(struct uio *uio)
+{
+	uma_zfree(uio_zone, uio);
+}
+
+static int
+uio_ctor(void *mem, int size, void *arg, int flags)
```

Comment:
So, because we have centralized allocation and deallocation routines for this zone, I'd argue that the ctor/dtor should simply be inlined into allocuio() and freeuio(). In general we've been discouraging the use of UMA ctor/dtors because the indirect function call induces some performance overhead, and because they're not particularly useful (unlike init/fini). They're not used very much in FreeBSD anymore.

Reply:
Adopted this approach now, moved the functions around to be near `cloneuio()`, which is probably more readable..